model: Allow to enable and disable package managers in the configuration

mnonnenmacher opened the pull request:
Allow to enable and disable specific package managers in the analyzer
configuration. This can still be overridden by the respective command
line options of the analyzer command.

The refactoring in this change is also a preparation for adding more
package manager specific configuration options later on.

sschuberth commented on model/src/main/kotlin/config/AnalyzerConfiguration.kt:
```diff
@@ -36,10 +36,17 @@ data class AnalyzerConfiguration(
+     * disabled.
      */
-    val options: Map<String, Options>? = null,
+    val enablePackageManagersByDefault: Boolean = true,
```
Do we really need this new flag? I have a tendency to not introduce it to reduce complexity. On CLI, also all PMs are enabled by default without the possibility to change that.

mnonnenmacher replied:
It's there for the same reason as the `-n` command line parameter. If you want to enable only one package manager, you would otherwise have to disable all others.

sschuberth replied:
I believe on the CLI being able to negate the logic made sense, as listing e.g. all but one package manager clutters the command line, and having to type the list of PMs over and over again is inconvenient. However, with the PM options in `ort.conf` I don't think that's a real issue anymore. You could simply search & replace if you'd want to "mass toggle" PMs.

That's why I'd still be in favor of removing that option completely. However, maybe there's an alternative: I know you already mentioned that you like your way of per-PM options. But many we could extract the `enabled` flag out of it, and still have my proposed lists of enabled / disabled PMs in `ort.conf`. This would be similar to e.g. the scanner storage configuration, where the storages themselves have their options, but whether they are actually used as readers / writers is configured separately.

sschuberth replied:
> I don't like the way it's implemented for scan storages either, because the configuration is scattered around.

It's maybe a question of encapsulation: Should a storage / PM itself know whether it's enabled, or is that the responsibility of an outer "manager"? I tend to say it's the latter.

> I would actually prefer if there were `read` and `write` flags in the storage configuration itself.

Even if getting a bit off topic: How would you then define the priority of storages?

Actually that question is semi-related to PMs: We'll soon have the need to configure the (relative) order in which PMs are run. I believe that requires a high-order concept anyway, similar to the mentioned enabled / disabled lists. Did you think about a syntax for that in `ort.conf` yet?

mnonnenmacher replied:
> It's maybe a question of encapsulation: Should a storage / PM itself know whether it's enabled, or is that the responsibility of an outer "manager"? I tend to say it's the latter.

I think the focus for the config file should be usability, not software design patterns, and how the values of the configuration classes are mapped to other classes is an implementation detail. Also for a value like `licenseFilePatterns` in the top-level configuration the location does not tell anything about in which classes this value is used internally, only that it is a global configuration that is not specific to one ORT tool.

> > I would actually prefer if there were `read` and `write` flags in the storage configuration itself.
> 
> Even if getting a bit off topic: How would you then define the priority of storages?

I would say the order in which they are defined. That would work because the priority is only relevant for reading, not for writing.

> Actually that question is semi-related to PMs: We'll soon have the need to configure the (relative) order in which PMs are run. I believe that requires a high-order concept anyway, similar to the mentioned enabled / disabled lists. Did you think about a syntax for that in `ort.conf` yet?

What we discussed separately was the we will need a way to configure that a package manager must run after (or before) another, so it would be like:
```
packageManagers {
  Gradle.mustRunAfter = Maven, NPM
}
```
Or in a more complete example it could look like:
```
enablePackageManagersByDefault = false

packageManagers {
  NPM.enabled = true

  NuGet {
    enabled = true
    mustRunAfter = NPM
    options {
      directDependenciesOnly = true
    }
  }
}
```
Or alternatively a `mustRunBefore` option. From the functionality they would be equivalent, so I'm not sure if we should have both or only one of them, and then which.

fviernau replied:
```
enablePackageManagersByDefault = false

packageManagers {
  NPM.enabled = true

  NuGet {
    enabled = true
    mustRunAfter = NPM
    options {
      directDependenciesOnly = true
    }
  }
}
``` 

Maybe one thing which hasn't been considered yet in above more complete example. If a package manager has configuration attributes without default values, then disabling a package manager (assuming `enablePackageManagersByDefault = true`) requires adding further configuration attributes. Probably this case isn't relevant as there are always default values, just wanted to bring it to consideration.

Apart from that we could add one more alternative implementation:
´´´
enabledPackageManagers = NPM, Yarn, Gradle
disabledPackageManager = GoMod
``` 
1. No default values for each package managers enabled state
2. Behavior: in case enabledPackageManagers + disabledPackageManager != allPackageManager: stop with an error sayinfg that the enabled state must be defined
3. This will enforce the user to make a conscious decision whether to enable any package manager after an upgrade of ORT which added a package manager

mnonnenmacher replied:
> Maybe one thing which hasn't been considered yet in above more complete example. If a package manager has configuration attributes without default values, then disabling a package manager (assuming `enablePackageManagersByDefault = true`) requires adding further configuration attributes. Probably this case isn't relevant as there are always default values, just wanted to bring it to consideration.

I'm not sure if I understand that correctly, but maybe it's a misunderstanding. If the config contains a line like `NuGet.enabled = false` it does not mean that an instance of `NuGet` is created and therefore options without default values would have to be provided. The actual class that is instantiated is `PackageManagerConfiguration`.

> Apart from that we could add one more alternative implementation:
> 
> ```
> enabledPackageManagers = NPM, Yarn, Gradle
> disabledPackageManager = GoMod
> ```
> 
>     1. No default values for each package managers enabled state
> 
>     2. Behavior: in case enabledPackageManagers + disabledPackageManager != allPackageManager: stop with an error saying that the enabled state must be defined
> 
>     3. This will enforce the user to make a conscious decision whether to enable any package manager after an upgrade of ORT which added a package manager

That's a good point, but as my intention is to allow this configuration also in the `.ort.yml` in a later step, it would not be manageable.

mnonnenmacher commented on cli/src/main/kotlin/commands/AnalyzerCommand.kt:
```diff
@@ -152,20 +154,22 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
         allPackageManagersByName[name]
             ?: throw BadParameterValue("Package managers must be one or more of ${allPackageManagersByName.keys}.")
-    }.split(",").default(PackageManager.ALL.toList())
+    }.split(",")
```
I think that's ok as previously the list of package managers was printed twice, so it's still printed once now.

mnonnenmacher replied:
The technical reason why I removed the default value is that I could not find a way to find out if the value was provided on the command line or not, but this is a requirement to decide if the values from the configuration file or from the command line should be used.

sschuberth replied:
For the record, such a check could probably be done via something like

    listOf("--package-managers", "-m").any { it in currentContext.originalArgv }

[but that's not really nice as it repeats the option names](https://github.com/ajalt/clikt/issues/359).

sschuberth commented on model/src/main/resources/reference.conf:
```diff
@@ -15,12 +15,18 @@ ort {
 
   analyzer {
     allowDynamicVersions = true
+    enablePackageManagersByDefault = true
```
I'm still not convinced that having this feature flag is a good idea, but I won't block it either. However, I'd like us to come up with a better name, as with `enablePackageManagersByDefault = false` it's not clear enough for *me* that it means *all* PMs are disabled by default.

It would be much easier with an *enum*, like `defaultPackageManagerState = {enabled,disabled}`, but if we'd like to stick with a *boolean*, maybe `disableAllPackageManagers` is better, as it both inverts the logic and explicitly mentions "All"?

mnonnenmacher replied:
I'm fine with the suggestions but I also like the one I proposed.
@oss-review-toolkit/core-devs As this is a breaking configuration change it would be nice to hear another opinion as well (to the change in general, not just the option name).

mnonnenmacher replied:
So I thought about the name again, if there is an option called `enablePackageManagersByDefault` and it is set to `false`, I don't understand why a user should not assume that this means that all package managers are disabled if not explicitly enabled.

sschuberth replied:
> I don't understand why a user should not assume that this means that all package managers are disabled if not explicitly enabled.

As mentioned before, I'm mostly missing an explicit "all".

But also, what this option should emphasize is that the default of enabling all PMs gets changed to disabling all PMs. So the emphasis is on "disable". That's why I'd prefer to also have "disable" in the name.

But like I said, I won't block this and will discard my change request; the naming doesn't really matter too much if I'm still having problems with the feature per se. I'll approve today if there's no other approval / further comments.

mnonnenmacher commented:
I have updated the PR based on the feedback on Slack and removed the breaking change label, as the breaking config change will be done in another PR now.

sschuberth commented on cli/src/main/kotlin/commands/AnalyzerCommand.kt:
```diff
@@ -286,3 +296,10 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
     }
 }
+
+private fun AnalyzerConfiguration.computeEnabledPackageManagers(): Set<PackageManagerFactory> {
```
Nit: "compute" is maybe a bit pretentious. Maybe just one of these: "get", "determine".

mnonnenmacher replied:
Used `determine` because `get` would override the generated getter for the property with the same name.

sschuberth commented on model/src/main/resources/reference.conf:
```diff
@@ -16,6 +16,13 @@ ort {
   analyzer {
     allowDynamicVersions = true
 
+    # A list of disabled package managers. Disabling a package manager here overrides enabling it in
+    # `enabledPackageManagers`.
+    disabledPackageManagers = [Maven, NPM]
```
Nit: Due to precedence explained in the comment, should we maybe make `disabledPackageManagers` go after `enabledPackageManagers` to emphasize that "the last word counts"?

sschuberth commented on cli/src/main/kotlin/commands/AnalyzerCommand.kt:
```diff
@@ -159,7 +160,10 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
             ?: throw BadParameterValue("Package managers must be one or more of ${allPackageManagersByName.keys}.")
-    }.split(",")
+    }.split(",").deprecated(
+        message = "--package-managers is deprecated, use -P ort.analyzer.enabledPackageManagers=[] instead.",
```
Is `[]` correct here? Don't we expect a comma-separated list without surrounding `[]` on the command line?

sschuberth commented on cli/src/main/kotlin/commands/AnalyzerCommand.kt:
```diff
@@ -159,7 +160,10 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
-    }.split(",")
+    }.split(",").deprecated(
+        message = "--package-managers is deprecated, use -P ort.analyzer.enabledPackageManagers=[] instead.",
+        tagValue = "use -P ort.analyzer.enabledPackageManagers=[] instead."
```
I haven't checked, but should this sentence start capitalized to look good in the CLI help?

mnonnenmacher replied:
This it what it looks like:
```
  -m, --package-managers VALUE     The comma-separated package managers to
                                   activate, any of [Bower, Bundler, Cargo,
                                   Carthage, CocoaPods, Composer, Conan,
                                   DotNet, GoDep, GoMod, Gradle, Maven, NPM,
                                   NuGet, PIP, Pipenv, Pub, SBT,
                                   SpdxDocumentFile, Stack, Unmanaged, Yarn,
                                   Yarn2]. Note that deactivation overrides
                                   activation. If set, package manager
                                   activation settings from configuration
                                   files are ignored. (deprecated: use -P
                                   ort.analyzer.enabledPackageManagers=[] on
                                   the ort command instead)
```

sschuberth commented on cli/src/main/kotlin/commands/AnalyzerCommand.kt:
```diff
@@ -159,7 +160,11 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
+    }.split(",").deprecated(
+        message = "--package-managers is deprecated, use -P ort.analyzer.enabledPackageManagers on the ort command " +
+                "instead.",
+        tagValue = "use -P ort.analyzer.enabledPackageManagers=[] on the ort command instead"
```
This still has `=[]`. Also, I wonder whether instead of dropping it completely, we should write something like `=...` (I haven't checked how we might already be doing this in other places).

sschuberth commented on cli/src/main/kotlin/commands/AnalyzerCommand.kt:
```diff
@@ -169,7 +174,11 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
+    }.split(",").deprecated(
+        message = "--not-package-managers is deprecated, use -P ort.analyzer.disabledPackageManagers on the ort " +
+                "command instead.",
+        tagValue = "use -P ort.analyzer.disabledPackageManagers=[] on the ort command instead"
```
See above.
